src/CurveController.sol

```diff
@@ -0,0 +1,243 @@
+/// SPDX-License-Identifier: GPL-3.0-or-later
+pragma solidity ^0.8.0;
+
+import "./interface/ISwapRouter.sol";
+import "./interface/IConvexRewards.sol";
+import "./interface/IPositionHandler.sol";
+import {ICurveController} from "./interface/ICurveController.sol";
+
+import "./library/Math.sol";
+
+import {SafeERC20} from "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+
+contract CurveController is ICurveController {
+    using SafeERC20 for IERC20;
+    using SafeERC20 for IERC20Metadata;
+
+    ISwapRouter public swapRouter;
+    // 0x3Fe65692bfCD0e6CF84cB1E7d24108E434A7587e
+    IConvexRewards public baseRewardPool;
+
+    address public governance;
+
+    constructor(
+        ISwapRouter _swapRouter,
+        IConvexRewards _baseRewardPool,
+        address _governance
+    ) {
+        swapRouter = _swapRouter;
+        baseRewardPool = _baseRewardPool;
+
+        governance = _governance;
+    }
+
+    function openPosition(
+        uint256 _amount,
+        bool _isLong,
+        uint256 _slippage,
+        bytes memory _data
+    ) external {
+        require(_isLong, "CurveController :: not long");
+        require(
+            _amount > 0 &&
+                _amount <= swapRouter.USDC().balanceOf(address(this)),
+            "CurveController :: amount"
+        );
+
+        /// Convert USDC -> CRV on 1inch
+        _safeApproveIfNotApproved(swapRouter.USDC(), address(swapRouter));
+        uint256 receivedCRV = swapRouter.estimateAndSwapTokens(
+            true,
+            address(swapRouter.CRV()),
+            _amount,
+            address(this),
+            _slippage,
+            _data
+        );
+
+        /// Convert CRV -> cvxCRV on Curve
+        _safeApproveIfNotApproved(swapRouter.CRV(), address(swapRouter));
+        swapRouter.swapOnCRVCVXCRVPool(true, receivedCRV, address(this));
+
+        /// Stake all cvxCRV on convex
+        _safeApproveIfNotApproved(swapRouter.CVXCRV(), address(baseRewardPool));
```

Comment:
Isn't it better to approve staker during initialisation itself to avoid check every time